src/location.h

```diff
@@ -0,0 +1,248 @@
+    }
+    if (valid_values_mask_ & HAS_SPEED) {
+      const uint16_t speedx100mps = speed_mps_ * 100.;
+      s.append(reinterpret_cast<const char *>(&speedx100mps), sizeof(speedx100mps));
+    }
+    return s;
+  }
+
+  // Initializes location from serialized byte array created by ToString() method.
+  // TODO(AlexZ): we don't care about endianness right now.
+  explicit Location(const std::string &encoded) { Decode(encoded); }
+
+  void Decode(const std::string &encoded) {
+    if (encoded.empty()) {
+      return;
+    }
+    std::string::size_type i = 0;
+    valid_values_mask_ = static_cast<Mask>(encoded[i++]);
+    if (valid_values_mask_ == NOT_INITIALIZED) return;
+    if (valid_values_mask_ & HAS_LATLON) {
+      timestamp_ms_ = *reinterpret_cast<const uint32_t *>(&encoded[i]) |
+                      (static_cast<uint64_t>(*reinterpret_cast<const uint16_t *>(&encoded[i + 4])) << 32);
+      i += sizeof(uint64_t) - 2;  // We use 6 bytes to store timestamps.
+      latitude_deg_ = *reinterpret_cast<const int32_t *>(&encoded[i]) / 10000000.;
+      i += sizeof(int32_t);
+      longitude_deg_ = *reinterpret_cast<const int32_t *>(&encoded[i]) / 10000000.;
+      i += sizeof(int32_t);
+      horizontal_accuracy_m_ = *reinterpret_cast<const uint32_t *>(&encoded[i]) / 100.;
+      i += sizeof(uint32_t);
+      if (valid_values_mask_ & HAS_SOURCE) {
+        source_ = static_cast<Source>(encoded[i++]);
+      }
+    }
+    if (valid_values_mask_ & HAS_ALTITUDE) {
+      altitude_m_ = *reinterpret_cast<const int32_t *>(&encoded[i]) / 100.;
+      i += sizeof(int32_t);
+      vertical_accuracy_m_ = *reinterpret_cast<const uint16_t *>(&encoded[i]) / 100.;
+      i += sizeof(uint16_t);
+    }
+    if (valid_values_mask_ & HAS_BEARING) {
+      bearing_deg_ = *reinterpret_cast<const uint32_t *>(&encoded[i]) / 10000000.;
+      i += sizeof(uint32_t);
+    }
+    if (valid_values_mask_ & HAS_SPEED) {
+      speed_mps_ = *reinterpret_cast<const uint16_t *>(&encoded[i]) / 100.;
+      i += sizeof(uint16_t);
+    }
+  }
+
+  Location() = default;
+
+  enum Source : std::uint8_t { UNKNOWN = 0, GPS = 1, NETWORK = 2, PASSIVE = 3 } source_;
+
+  bool HasLatLon() const { return valid_values_mask_ & HAS_LATLON; }
+  Location &SetLatLon(uint64_t timestamp_ms,
+                      double latitude_deg,
+                      double longitude_deg,
+                      double horizontal_accuracy_m) {
+    if (horizontal_accuracy_m > 0.0) {
+      timestamp_ms_ = timestamp_ms;
+      latitude_deg_ = latitude_deg;
+      longitude_deg_ = longitude_deg;
+      horizontal_accuracy_m_ = horizontal_accuracy_m;
+      valid_values_mask_ = static_cast<Mask>(valid_values_mask_ | HAS_LATLON);
+    }
+    return *this;
+  }
+
+  bool HasSource() const { return valid_values_mask_ & HAS_SOURCE; }
+  Location &SetSource(Source source) {
+    source_ = source;
+    valid_values_mask_ = static_cast<Mask>(valid_values_mask_ | HAS_SOURCE);
```

Comment:
It does not work with enums. I prefer it in this way for easier readability - you can see exactly what's going on.